### Importing relevant libraries

In [1]:
import pandas as pd
import numpy as np

#### loading data

In [2]:
data_preprocessed = pd.read_csv('../Absenteeism_preprocessed.csv')

In [3]:
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Date of the Week  \
0         0         0         0         1            7                 1   
1         0         0         0         0            7                 1   
2         0         0         0         1            7                 2   
3         1         0         0         0            7                 3   
4         0         0         0         1            7                 3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

#### Creating targets

3.0

Calculated the median, anything below the median is ok, anything above the median is excessive Absenteeism 

In [7]:
targets = np.where(data_preprocessed['Absenteeism Time in Hours'] > data_preprocessed['Absenteeism Time in Hours'].median(), 1 , 0)

In [8]:
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [9]:
data_preprocessed['Excessive Absenteeism'] = targets

In [10]:
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Date of the Week  \
0         0         0         0         1            7                 1   
1         0         0         0         0            7                 1   
2         0         0         0         1            7                 2   
3         1         0         0         0            7                 3   
4         0         0         0         1            7                 3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0               30          0         2     1                          4   
1               31          0         1     0                          0   
2               31          0         0     0                          2   
3               24          0         2     0                          4   
4               30          0         2     1                          2   

   Excessive Absenteeism  
0                      1  
1                      0  
2                      0  
3                      1  
4                      0

#### A comment on targets

In [11]:
targets.sum() / targets.shape[0]

0.45571428571428574

In [12]:
data_with_targets = data_preprocessed.drop(['Absenteeism Time in Hours'],axis = 1)

Checking if the 2 datasets have the same place in the memory (true/false)

In [13]:
data_with_targets is data_preprocessed

False

In [14]:
data_with_targets.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Date of the Week  \
0         0         0         0         1            7                 1   
1         0         0         0         0            7                 1   
2         0         0         0         1            7                 2   
3         1         0         0         0            7                 3   
4         0         0         0         1            7                 3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0               30          0         2     1                      1  
1               31          0         1     0                      0  
2               31          0         0     0                      0  
3               24          0         2     0                      1  
4               30          0         2     1                      0

### Selecting inputs for the regression

In [16]:
data_with_targets.shape

(700, 15)

In [18]:
data_with_targets.iloc[:,0:-1]

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Date of the Week  \
0           0         0         0         1            7                 1   
1           0         0         0         0            7                 1   
2           0         0         0         1            7                 2   
3           1         0         0         0            7                 3   
4           0         0         0         1            7                 3   
..        ...       ...       ...       ...          ...               ...   
695         1         0         0         0            5                 2   
696         1         0         0         0            5                 2   
697         1         0         0         0            5                 3   
698         0         0         0         1            5                 3   
699         0         0         0         1            5                 3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  
0                 30          0         2     1  
1                 31          0         1     0  
2                 31          0         0     0  
3                 24          0         2     0  
4                 30          0         2     1  
..               ...        ...       ...   ...  
695               22          1         2     0  
696               24          0         1     2  
697               25          1         0     0  
698               25          1         0     0  
699               25          0         1     1  

[700 rows x 14 columns]

In [19]:
unscaled_inputs = data_with_targets.iloc[:,0:-1]

#### Standrize the Data

In [23]:
from sklearn.preprocessing import StandardScaler
absenteeism_scaler = StandardScaler()

In [24]:
absenteeism_scaler.fit(unscaled_inputs)

StandardScaler()

In [25]:
scaled_inputs = absenteeism_scaler.transform(unscaled_inputs)

In [26]:
scaled_inputs

array([[-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         0.88046927,  0.26848661],
       [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
        -0.01928035, -0.58968976],
       [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
        -0.91902997, -0.58968976],
       ...,
       [ 1.73205081, -0.09298136, -0.31722063, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31722063, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
        -0.01928035,  0.26848661]])

In [27]:
scaled_inputs.shape

(700, 14)

### Split the data into train & test and shuffle

#### Import the relevant module

In [28]:
from sklearn.model_selection import train_test_split

#### Split

In [38]:
train_test_split(scaled_inputs,targets)

[array([[-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         -0.91902997, -0.58968976],
        [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         -0.01928035,  1.12666297],
        [-0.57735027, -0.09298136, -0.31722063, ...,  2.23224237,
         -0.91902997, -0.58968976],
        ...,
        [-0.57735027, -0.09298136, -0.31722063, ...,  2.23224237,
         -0.91902997, -0.58968976],
        [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         -0.91902997, -0.58968976],
        [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         -0.01928035,  0.26848661]]),
 array([[-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
          2.67996851, -0.58968976],
        [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
         -0.01928035,  1.12666297],
        [-0.57735027, -0.09298136, -0.31722063, ..., -0.44798003,
          0.88046927,  0.26848661],
        ...,
        [ 1.73205081, -0.09298136, -0.31722063, ..., -

In [40]:
x_train, x_test, y_train, y_test = train_test_split(scaled_inputs,targets, train_size= 0.8, random_state= 20)

In [41]:
print(x_train.shape, y_train.shape)

(560, 14) (560,)


The above result means the inputs have 525 observations along 14 features, while the targets are a vector of length 525

In [32]:
print(x_test.shape, y_test.shape)

(175, 14) (175,)


The test includes 175 observations, 14 input variables and 1 target variable

### Logistic Regression with Sklearn

In [42]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

#### Training the model

In [43]:
reg = LogisticRegression()

In [45]:
reg.fit(x_train, y_train)
print(reg)

LogisticRegression()


In [48]:
reg.score(x_train,y_train)

0.7821428571428571

Based on the data we used, the model was able to classify around 80% of the observations correctly

#### Manually checking the accuracy

In [50]:
model_outputs = reg.predict(x_train)
model_outputs

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,

In [51]:
y_train

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,

In [52]:
model_outputs == y_train

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True, False, False,  True,  True,  True,  True,
       False,  True, False,  True, False, False,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False, False, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,  True,  True, False,  True, False,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
       False,  True, False,  True,  True, False, False, False,  True,
        True,  True,  True,  True,  True,  True,  True, False,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False,  True,  True,  True,  True,
       False,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,

In [53]:
np.sum((model_outputs==y_train))

438

In [54]:
model_outputs.shape[0]

560

In [56]:
# Accuracy = Correct predictions/ #observations
np.sum((model_outputs==y_train)) / model_outputs.shape[0]

0.7821428571428571

This output means that our result is exactly the same from the sklearn method score

#### Finding the Intercept and Coefficients

In [58]:
reg.intercept_

array([-0.21926128])

In [59]:
reg.coef_

array([[ 2.08531551,  0.27632288,  1.58266848,  1.33318134,  0.18995161,
        -0.06485041,  0.71589014, -0.04392642, -0.20539315, -0.02264482,
         0.31997743, -0.14797072,  0.37645987, -0.33760641]])

In [60]:
unscaled_inputs.columns.values

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Date of the Week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [64]:
feature_name= unscaled_inputs.columns.values
feature_name

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Date of the Week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [65]:
summary_table = pd.DataFrame(columns=['Feature name'], data=feature_name)
summary_table['Coefficient'] = np.transpose(reg.coef_)
summary_table

Feature name  Coefficient
0                  Reason_1     2.085316
1                  Reason_2     0.276323
2                  Reason_3     1.582668
3                  Reason_4     1.333181
4               Month Value     0.189952
5          Date of the Week    -0.064850
6    Transportation Expense     0.715890
7          Distance to Work    -0.043926
8                       Age    -0.205393
9   Daily Work Load Average    -0.022645
10          Body Mass Index     0.319977
11                Education    -0.147971
12                 Children     0.376460
13                     Pets    -0.337606

In [66]:
summary_table.index = summary_table.index+1
summary_table.loc[0]= ['intercept', reg.intercept_[0]]
summary_table = summary_table.sort_index()
summary_table

Feature name  Coefficient
0                 intercept    -0.219261
1                  Reason_1     2.085316
2                  Reason_2     0.276323
3                  Reason_3     1.582668
4                  Reason_4     1.333181
5               Month Value     0.189952
6          Date of the Week    -0.064850
7    Transportation Expense     0.715890
8          Distance to Work    -0.043926
9                       Age    -0.205393
10  Daily Work Load Average    -0.022645
11          Body Mass Index     0.319977
12                Education    -0.147971
13                 Children     0.376460
14                     Pets    -0.337606